lib/block_view_stake.go

```diff
@@ -0,0 +1,2600 @@
+		)
+	}
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*UnlockStakeMetadata)
+
+	// Validate the txn metadata.
+	if err := bav.IsValidUnlockStakeMetadata(txn.PublicKey, txMeta); err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: ")
+	}
+
+	// Convert TransactorPublicKey to TransactorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return 0, 0, nil, errors.Wrapf(RuleErrorInvalidStakerPKID, "_connectUnlockStake: ")
+	}
+
+	// Convert ValidatorPublicKey to ValidatorPKID.
+	validatorPKIDEntry := bav.GetPKIDForPublicKey(txMeta.ValidatorPublicKey.ToBytes())
+	if validatorPKIDEntry == nil || validatorPKIDEntry.isDeleted {
+		return 0, 0, nil, errors.Wrapf(RuleErrorInvalidValidatorPKID, "_connectUnlockStake: ")
+	}
+
+	// Retrieve the PrevLockedStakeEntries. These will be restored if we disconnect this txn.
+	prevLockedStakeEntries, err := bav.GetLockedStakeEntriesInRange(
+		validatorPKIDEntry.PKID, transactorPKIDEntry.PKID, txMeta.StartEpochNumber, txMeta.EndEpochNumber,
+	)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: ")
+	}
+	if len(prevLockedStakeEntries) == 0 {
+		return 0, 0, nil, errors.Wrapf(RuleErrorInvalidUnlockStakeNoUnlockableStakeFound, "_connectUnlockStake: ")
+	}
+
+	// Connect a basic transfer to get the total input and the
+	// total output without considering the txn metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures,
+	)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: ")
+	}
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the txn is signed
+		// by the top-level public key, which we take to be the sender's
+		// public key so there is no need to verify anything further.
+	}
+
+	// Calculate the TotalUnlockedAmountNanos and delete the PrevLockedStakeEntries.
+	totalUnlockedAmountNanos := uint256.NewInt()
+	for _, prevLockedStakeEntry := range prevLockedStakeEntries {
+		totalUnlockedAmountNanos, err = SafeUint256().Add(
+			totalUnlockedAmountNanos, prevLockedStakeEntry.LockedAmountNanos,
+		)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: ")
+		}
+		bav._deleteLockedStakeEntryMappings(prevLockedStakeEntry)
+	}
+	if !totalUnlockedAmountNanos.IsUint64() {
+		return 0, 0, nil, errors.Wrapf(RuleErrorInvalidUnlockStakeUnlockableStakeOverflowsUint64, "_connectUnlockStake: ")
+	}
+	totalUnlockedAmountNanosUint64 := totalUnlockedAmountNanos.Uint64()
+
+	// Add TotalUnlockedAmountNanos to TotalInput. The unlocked coins are an
+	// implicit input even though they do not come from a specific public key.
+	totalInput, err = SafeUint64().Add(totalInput, totalUnlockedAmountNanosUint64)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUnlockStake: error adding TotalUnlockedAmountNanos to TotalInput: ")
+	}
```

Comment:
As a side-note, in Rosetta we're going to have to ues the "sub-account" thing like we did with Creator Coins to handle this.

Reply:
No action on this yet.